### PR TITLE
feat(swap): auto-fallback to streaming swap when THORChain rapid slippage >3%

### LINKED ISSUE #4145: [ADD] Anti-rekt: auto-fallback to THORChain streaming swap when rapid slippage exceeds 3%
## Problem

THORChain rapid swaps (`interval=0`) on shallow pools can inflict catastrophic slippage that surfaces as `fees.total` in the quote. Observed in-app:

- **1 BTC → TRX**: input $75,766, output $24,079 → **$43,440 (57%) "Swap Fee"** from the new shallow TRON pool
- **1 BTC → XRP**: input $75,766, output $55,720 → **$12,504 (16%) "Swap Fee"**

Both numbers are legitimate `fees.total` values from the THORChain quote endpoint — the AMM's liquidity fee scales with trade-size vs pool-depth. Users have no in-wallet protection. A comparable `[ADD] Max slippage for Lifi/1inch swaps` was shipped in 2024 (#1408) for EVM providers; no equivalent exists for THORChain.

## Prior context

The rapid-vs-streaming decision has deliberate history:
- Android #3702 (closed) — THORChain `interval` switched from 1 → 0
- iOS #4088 (closed) — same
- Windows #3613 (open) — same switch still pending

The switch to rapid was **correct**: streaming swaps take multiple blocks (5-30+ minutes) and the "confirm → 20 minutes of silence" UX confused users. Rapid is right for ~99% of trades. **This issue does not reverse that decision — it adds an anti-rekt escape hatch** that keeps rapid as default and only reaches for streaming when the user would otherwise eat >3% slippage on a shallow pool.

## Proposed solution

In `SwapQuoteRepository.getSwapQuote` (THORChain branch only):

1. Fetch rapid quote as today (`interval=0`). — **Normal path. No behavior change for ~99% of trades.**
2. Compute `rapidSlippageBps = (fees.total_in_fiat / input_fiat) × 10_000`.
3. If `rapidSlippageBps ≤ 300` (3%): return rapid. Done.
4. Else fetch a streaming quote with `interval=1, streaming_quantity=rapid.max_streaming_quantity` (the quote response's own optimal suggestion).
5. If streaming fetch fails: log and return rapid (silent fallback — user sees current behavior, no new error surface).
6. If `streaming.expected_amount_out > rapid.expected_amount_out`: return streaming. Else: rapid.

The `memo` field in the quote response already embeds `/INTERVAL/QUANTITY` when streaming params are requested. **Signing, keysign, and UTXO `OP_RETURN` code remains unchanged** — verified that `quote.data.memo` passes through verbatim in Android/iOS/Windows.

## Scope

**In scope**
- Threshold-based double-fetch + quote selection in `SwapQuoteRepository`.
- Single configurable constant `STREAMING_SLIPPAGE_THRESHOLD_BPS = 300`.
- Unit tests for the four branches (below threshold, above + streaming better, above + streaming worse, streaming fetch error).

**Out of scope (explicit)**
- **No UI changes.** No rapid/streaming toggle, no ETA label, no streaming badge. Users always see the best available quote. Execution mode is observable via tx history / on-chain.
- No time cap on streaming duration — trust THORChain's `max_streaming_quantity`.
- Maya provider (same quote shape; separate ticket if we want parity).
- Telemetry (follow-up; useful for tuning but not blocking).

## Acceptance

- [ ] Rapid `fees.total ≤ 3%` → streaming quote NOT fetched (normal path unchanged)
- [ ] Rapid `fees.total > 3%` → streaming quote fetched
- [ ] Streaming fetch exception → rapid returned, no exception to UI
- [ ] `streaming.expected_amount_out > rapid` → streaming memo (`/INTERVAL/QUANTITY`) surfaces to signing
- [ ] `streaming.expected_amount_out ≤ rapid` → rapid returned
- [ ] Unit tests for all four branches
- [ ] Manual QA: **1 BTC → TRX** auto-switches to streaming with meaningful output improvement; **0.01 BTC → ETH** stays rapid unchanged

## Risk / Rollback

- One constant. Set to `Int.MAX_VALUE` to fully disable without reverting code.
- At most one extra THORChain API call per quote refresh, only when slippage > 3%. Bounded by the existing 450ms debounce.
- If THORChain public endpoint rate-limits: streaming fetch fails → silent fallback → rapid returned → user sees current behavior.
- Rollback: single `git revert`. No schema, no persisted state, no migration.

## Reference (Android)

- Quote fetch: `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt` (`getSwapQuote` — THORChain branch)
- Quote model: `data/src/main/kotlin/com/vultisig/wallet/data/api/models/quotes/THORChainSwapQuote.kt` — already exposes `expected_amount_out`, `max_streaming_quantity`, `fees.total`, `memo`
- Fiat conversion helpers: `app/src/main/java/com/vultisig/wallet/ui/models/swap/SwapQuoteManager.kt`
- Memo pass-through (verified unchanged): `SwapFormViewModel.kt`, `ThorChainHelper.kt`, `ThorchainSwapHelper.kt`

## Platform parity

Sibling tickets to be opened on `vultisig/vultisig-ios` and `vultisig/vultisig-windows` with the same spec. Signing-side memo handling is already consistent across all three platforms, so only the quote-fetch layer changes per repo.

### PR DESCRIPTION
**Pull request by johnnyluo**
## Summary

- Adds a slippage guard for THORChain rapid swaps (`interval=0`)
- When `fees.total / gross_output > 3%` (300 bps), fetches a streaming quote (`interval=1`, `streaming_quantity=max_streaming_quantity`) and returns whichever gives better `expected_amount_out`
- Streaming fetch failures fall back silently to rapid — no new error surface, no regression for the ~99% of swaps that stay on the rapid path
- Single tuning constant `STREAMING_SLIPPAGE_THRESHOLD_BPS = 300`; set to `Int.MAX_VALUE` to disable without reverting

## Changes

- `ThorChainSwapQuoteRequest` — adds optional `streamingQuantity` field (default `null`)
- `ThorChainApi.getSwapQuotes` — passes `streaming_quantity` query param when provided
- `SwapQuoteRepository.getSwapQuote` — implements the double-fetch + selection logic
- `SwapQuoteRepositoryThorChainStreamingTest` — 6 unit tests covering all four branches: below threshold, above + streaming better, above + streaming worse, streaming throws, streaming returns API error, streaming returns `error` field

## Acceptance criteria coverage

- [x] Rapid `fees.total ≤ 3%` → streaming NOT fetched
- [x] Rapid `fees.total > 3%` → streaming fetched
- [x] Streaming fetch exception → rapid returned, no exception to UI
- [x] `streaming.expected_amount_out > rapid` → streaming memo surfaces to signing
- [x] `streaming.expected_amount_out ≤ rapid` → rapid returned
- [x] Unit tests for all branches

Closes #4145

## Test plan

- [ ] Run `./gradlew :data:testDebugUnitTest` — all 6 new tests pass
- [ ] Manual QA: **1 BTC → TRX** auto-switches to streaming with meaningful output improvement
- [ ] Manual QA: **0.01 BTC → ETH** (low slippage) stays on rapid, no second API call

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Optional streaming-quantity can be included with THORChain swap quote requests.
  * Quote selection now compares rapid vs streaming quotes, preferring the better expected output and applying a slippage threshold.
  * Streaming errors gracefully fall back to rapid; rapid failures trigger a streaming attempt.

* **Tests**
  * New test suite covering rapid vs streaming selection, fallback behavior, and failure modes.

### CODE REVIEW

File: data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt [hunk 1]
```diff
@@ -264,47 +266,85 @@ constructor(
...
-                    data = thorQuote.data,
-                    expiredAt = Clock.System.now() + expiredAfter,
-                )
+        val feesTotal = rapidData.fees.total.toBigInteger()
+        val rapidExpectedOut = rapidData.expectedAmountOut.toBigInteger()
+        val grossOut = feesTotal + rapidExpectedOut
+        val rapidSlippageBps =
+            if (grossOut > BigInteger.ZERO)
+                feesTotal.multiply(BigInteger.valueOf(10000)).divide(grossOut).toInt()
+            else 0
```
[the remark this answers is not in the document]
Comment (johnnyluo) [reply data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt]: that should be okay

File: data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt [hunk 2]
```diff
@@ -264,47 +267,107 @@ constructor(
...
+                interval = "0",
+                referralCode = referralCode,
+                bpsDiscount = bpsDiscount,
+            )
 
-        when (thorQuote) {
-            is THORChainSwapQuoteDeserialized.Error -> {
-                throw SwapException.handleSwapException(thorQuote.error.message)
+        // Fetch rapid quote; capture failure so we can fall back to streaming
+        var rapidData: THORChainSwapQuote? = null
```
Comment (aminsato) [inline data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:282]: _Suggestion (optional):_ the `var rapidData` / `var rapidError` pair forces the workaround `val rapidSnapshot = rapidData` at L327 and the two compiler-hint comments at L307 and L326. Consider extracting a small sealed result type and returning it from a helper, e.g.:

```kotlin
private sealed interface RapidQuoteResult {
    data class Success(val data: THORChainSwapQuote) : RapidQuoteResult
    data class Failure(val error: SwapException) : RapidQuoteResult
}

private suspend fun fetchRapidQuote(req: ThorChainSwapQuoteRequest): RapidQuoteResult = ...
```

The rest of the function then operates on immutable values and the smart-cast gymnastics disappear.

**Review by aminsato — COMMENTED** [review]
A few non-blocking suggestions on the THORChain rapid/streaming fallback — the core logic and tests look solid.

File: data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt [hunk 3]
```diff
@@ -428,6 +491,18 @@ constructor(
         )
     }
 
+    private fun pickBestQuote(
+        rapid: THORChainSwapQuote?,
+        streaming: THORChainSwapQuote?,
+        rapidError: SwapException?,
+    ): THORChainSwapQuote {
+        if (streaming == null && rapid == null) throw requireNotNull(rapidError)
```
Comment (aminsato) [inline data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:499]: _Suggestion (optional):_ `throw requireNotNull(rapidError)` reads as "assert non-null, then throw" but the intent is "surface the rapid error if both fetches failed". `throw rapidError ?: error("both rapid and streaming quote fetches failed")` is shorter and gives a useful message if the invariant is ever broken in the future.

File: data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt [hunk 4]
```diff
@@ -597,5 +672,6 @@ constructor(
     companion object {
         private const val SOLANA_DEFAULT_CONTRACT_ADDRESS =
             "So11111111111111111111111111111111111111112"
+        const val STREAMING_SLIPPAGE_THRESHOLD_BPS = 300
```
Comment (aminsato) [inline data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:675]: _Suggestion (optional):_ the sibling `SOLANA_DEFAULT_CONTRACT_ADDRESS` two lines above is `private const val`, but `STREAMING_SLIPPAGE_THRESHOLD_BPS` is exposed without the `private` modifier. It's only referenced inside this file, so marking it `private` keeps the companion's visibility consistent. If it's intentionally public for tuning/testing, a one-line KDoc explaining the choice would help.

**Review by premiumjibles — COMMENTED** [review]
*How do you do, fellow humans. I've had a look at your code.*

Adds a slippage guard around THORChain rapid swaps (interval=0): when fees.total / gross_output exceeds 300 bps, fetch a parallel streaming quote (interval=1, streaming_quantity=max_streaming_quantity) and return whichever has the better expected_amount_out. Streaming-fetch failures fall back silently to rapid; a new pickBestQuote helper plus 9 unit tests cover all rapid/streaming combinations.

### What's right

- Single tuning knob (STREAMING_SLIPPAGE_THRESHOLD_BPS) with documented disable-by-MAX_VALUE escape hatch — keeps the behaviour reversible without a code revert.
- Asymmetric fallback: rapid failure also triggers streaming, not just the slippage gate — covers the case where a pool is paused for rapid but streaming-acceptable.
- Tests exercise the four-quadrant matrix (rapid×streaming success/failure) plus the streaming-error-field branch the API distinguishes from HTTP errors.

### Findings

#### **[MAJOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:315`
_Slippage bps uses integer floor — values just above 3% gated out_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:315 — Also flagged by CodeRabbit. The bps formula `feesTotal.multiply(10000).divide(grossOut).toInt()` floors before the `> STREAMING_SLIPPAGE_THRESHOLD_BPS` comparison, so a swap whose true slippage is 300.99 bps reads as 300 and skips the streaming path. The PR contract is `> 3%` but the code implements `floor(bps) > 300` (i.e. ≥301 bps). Consider comparing without flooring (e.g. `feesTotal.multiply(10000) > grossOut.multiply(BigInteger.valueOf(STREAMING_SLIPPAGE_THRESHOLD_BPS.toLong()))`) so the gate matches the PR description.



#### **[MAJOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:360`
_Streaming quote expiry uses rapid's expiredAfter_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:360 — When the streaming quote wins, the returned SwapQuote.ThorChain is stamped with `Clock.System.now() + expiredAfter` (the rapid-path expiry constant), but a streaming swap with `streaming_quantity=N, interval=1` runs for N THORChain blocks (~6s each) plus inbound/outbound — easily several minutes. If `expiredAfter` was tuned for rapid (single-block) lifetimes, users may see the quote expire mid-flow. Consider deriving expiry from streaming.totalSwapSeconds when the streaming branch is taken.



#### **[MINOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:502`
_pickBestQuote ignores execution-time delta_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:502 — pickBestQuote picks streaming whenever streamingOut > rapidOut, even by a single base unit. A streaming swap that improves output by 0.001% but takes 5 minutes vs the rapid path's seconds is a UX regression. Consider requiring a meaningful delta (e.g. ≥1% improvement) before swapping, especially since the slippage gate already establishes that rapid is materially worse.



#### **[MINOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:336`
_Streaming retry on rapid API error masks pool-suspended / pair-invalid signals_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:336 — When rapid returns Result(error='…') or an HTTP error, the new code unconditionally tries streaming. Some rapid-side errors (`pool suspended`, `pair not supported`) will reproduce on streaming; others (`min amount in`) won't. Consider only triggering streaming fallback for transient/network errors and propagating structured API errors.



#### **[MINOR]** (confidence: high) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:328`
_streamingQuantity sent without lower-bound check_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:328 — `streamingQuantityHint = rd.maxStreamingQuantity` — if THORChain returns max_streaming_quantity = 0 for tiny inputs or paused pools, this PR sends `streaming_quantity=0` which the API will likely reject. Consider `rd.maxStreamingQuantity.takeIf { it > 0 }`.



#### **[QUESTION]** (confidence: low) — `data/src/test/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepositoryThorChainStreamingTest.kt:117`
_No test asserts streaming memo flows to SwapQuote.ThorChain_

data/src/test/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepositoryThorChainStreamingTest.kt:117 — Tests assert expectedAmountOut matches but don't assert returned data.memo is the streaming memo (rather than rapid's). A `SWAP:TRON.TRX:TDst:0/1/5` (streaming) vs `SWAP:TRON.TRX:TDst` (rapid) memo distinction is what gets signed downstream.



#### **[NIT]** (confidence: high) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:675`
_STREAMING_SLIPPAGE_THRESHOLD_BPS exposed as public companion const_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:675 — `const val STREAMING_SLIPPAGE_THRESHOLD_BPS = 300` in the companion is implicitly public. Consider `private` (or `internal`) so it doesn't accidentally become a load-bearing constant for other modules.



#### **[NIT]** (confidence: high) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:305`
_Rapid-failure log drops the underlying error_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:305 — `Timber.w("Rapid quote failed, trying streaming fallback")` discards rapidError. Consider `Timber.w(rapidError, "Rapid quote failed (%s); trying streaming fallback", rapidError?.message)`.



### QA Report

| Check | Result |
|---|---|
| build/lint | not_run — sweep skipped checkout |
| tests | not_run — CI reports build-and-test passed |
| runtime probe | not applicable — mobile (Android) |

### Dedup notes

- CodeRabbit: integer-floor on rapidSlippageBps comparison — file SwapQuoteRepository.kt around 312-317
- CodeRabbit: force-unwrap !! on quote.data.error inside null-checked branch — file SwapQuoteRepository.kt around 289-292
- CodeRabbit: missing exact-threshold test (300 bps boundary) — already on author's plate
- CodeRabbit: missing log/telemetry on non-exception streaming failures — already flagged

## Verdict

**COMMENT**

---
🤖 review-pr@Claude-4.7 — single-agent run (Gemini second-pass deferred for sweep budget). See ~/.claude/skills/review-pr if this got something wrong.

**Review by premiumjibles — COMMENTED** [review]
*How do you do, fellow humans. I've had a look at your code.*

Adds a slippage guard around THORChain rapid swaps (interval=0): when fees.total / gross_output exceeds 300 bps, fetch a parallel streaming quote (interval=1, streaming_quantity=max_streaming_quantity) and return whichever has the better expected_amount_out. Streaming-fetch failures fall back silently to rapid; a new pickBestQuote helper plus 9 unit tests cover all rapid/streaming combinations.

### What's right

- Single tuning knob (STREAMING_SLIPPAGE_THRESHOLD_BPS) with documented disable-by-MAX_VALUE escape hatch — keeps the behaviour reversible without a code revert.
- Asymmetric fallback: rapid failure also triggers streaming, not just the slippage gate — covers the case where a pool is paused for rapid but streaming-acceptable.
- Tests exercise the four-quadrant matrix (rapid×streaming success/failure) plus the streaming-error-field branch the API distinguishes from HTTP errors.

### Findings

#### **[MAJOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:315`
_Slippage bps uses integer floor — values just above 3% gated out_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:315 — Also flagged by CodeRabbit. The bps formula `feesTotal.multiply(10000).divide(grossOut).toInt()` floors before the `> STREAMING_SLIPPAGE_THRESHOLD_BPS` comparison, so a swap whose true slippage is 300.99 bps reads as 300 and skips the streaming path. The PR contract is `> 3%` but the code implements `floor(bps) > 300` (i.e. ≥301 bps). Consider comparing without flooring (e.g. `feesTotal.multiply(10000) > grossOut.multiply(BigInteger.valueOf(STREAMING_SLIPPAGE_THRESHOLD_BPS.toLong()))`) so the gate matches the PR description.



#### **[MAJOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:360`
_Streaming quote expiry uses rapid's expiredAfter_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:360 — When the streaming quote wins, the returned SwapQuote.ThorChain is stamped with `Clock.System.now() + expiredAfter` (the rapid-path expiry constant), but a streaming swap with `streaming_quantity=N, interval=1` runs for N THORChain blocks (~6s each) plus inbound/outbound — easily several minutes. If `expiredAfter` was tuned for rapid (single-block) lifetimes, users may see the quote expire mid-flow. Consider deriving expiry from streaming.totalSwapSeconds when the streaming branch is taken.



#### **[MINOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:502`
_pickBestQuote ignores execution-time delta_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:502 — pickBestQuote picks streaming whenever streamingOut > rapidOut, even by a single base unit. A streaming swap that improves output by 0.001% but takes 5 minutes vs the rapid path's seconds is a UX regression. Consider requiring a meaningful delta (e.g. ≥1% improvement) before swapping, especially since the slippage gate already establishes that rapid is materially worse.



#### **[MINOR]** (confidence: med) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:336`
_Streaming retry on rapid API error masks pool-suspended / pair-invalid signals_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:336 — When rapid returns Result(error='…') or an HTTP error, the new code unconditionally tries streaming. Some rapid-side errors (`pool suspended`, `pair not supported`) will reproduce on streaming; others (`min amount in`) won't. Consider only triggering streaming fallback for transient/network errors and propagating structured API errors.



#### **[MINOR]** (confidence: high) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:328`
_streamingQuantity sent without lower-bound check_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:328 — `streamingQuantityHint = rd.maxStreamingQuantity` — if THORChain returns max_streaming_quantity = 0 for tiny inputs or paused pools, this PR sends `streaming_quantity=0` which the API will likely reject. Consider `rd.maxStreamingQuantity.takeIf { it > 0 }`.



#### **[QUESTION]** (confidence: low) — `data/src/test/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepositoryThorChainStreamingTest.kt:117`
_No test asserts streaming memo flows to SwapQuote.ThorChain_

data/src/test/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepositoryThorChainStreamingTest.kt:117 — Tests assert expectedAmountOut matches but don't assert returned data.memo is the streaming memo (rather than rapid's). A `SWAP:TRON.TRX:TDst:0/1/5` (streaming) vs `SWAP:TRON.TRX:TDst` (rapid) memo distinction is what gets signed downstream.



#### **[NIT]** (confidence: high) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:675`
_STREAMING_SLIPPAGE_THRESHOLD_BPS exposed as public companion const_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:675 — `const val STREAMING_SLIPPAGE_THRESHOLD_BPS = 300` in the companion is implicitly public. Consider `private` (or `internal`) so it doesn't accidentally become a load-bearing constant for other modules.



#### **[NIT]** (confidence: high) — `data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:305`
_Rapid-failure log drops the underlying error_

data/src/main/kotlin/com/vultisig/wallet/data/repositories/SwapQuoteRepository.kt:305 — `Timber.w("Rapid quote failed, trying streaming fallback")` discards rapidError. Consider `Timber.w(rapidError, "Rapid quote failed (%s); trying streaming fallback", rapidError?.message)`.



### QA Report

| Check | Result |
|---|---|
| build/lint | not_run — sweep skipped checkout |
| tests | not_run — CI reports build-and-test passed |
| runtime probe | not applicable — mobile (Android) |

### Dedup notes

- CodeRabbit: integer-floor on rapidSlippageBps comparison — file SwapQuoteRepository.kt around 312-317
- CodeRabbit: force-unwrap !! on quote.data.error inside null-checked branch — file SwapQuoteRepository.kt around 289-292
- CodeRabbit: missing exact-threshold test (300 bps boundary) — already on author's plate
- CodeRabbit: missing log/telemetry on non-exception streaming failures — already flagged

## Verdict

**COMMENT**

---
🤖 review-pr@Claude-4.7 — single-agent run (Gemini second-pass deferred for sweep budget). See ~/.claude/skills/review-pr if this got something wrong.
